src/System.Private.ServiceModel/src/System/ServiceModel/Channels/BufferedStream.cs

```diff
@@ -0,0 +1,1098 @@
+    /// This class will never cache more bytes than the max specified buffer size.
+    /// However, it may use a temporary buffer of up to twice the size in order to combine several IO operations on
+    /// the underlying stream into a single operation. This is because we assume that memory copies are significantly
+    /// faster than IO operations on the underlying stream (if this was not true, using buffering is never appropriate).
+    /// The max size of this "shadow" buffer is limited as to not allocate it on the LOH.
+    /// Shadowing is always transient. Even when using this technique, this class still guarantees that the number of
+    /// bytes cached (not yet written to the target stream or not yet consumed by the user) is never larger than the 
+    /// actual specified buffer size.
+    /// </summary>
+    internal sealed class BufferedStream : Stream
+    {
+        private const int DefaultBufferSize = 4096;
+
+
+        private Stream _stream; // Underlying stream.  Close sets _stream to null.
+
+        private byte[] _buffer; // Shared read/write buffer.  Alloc on first use.
+
+        private readonly int _bufferSize; // Length of internal buffer (not counting the shadow buffer).
+
+        private int _readPos; // Read pointer within shared buffer.
+        private int _readLen; // Number of bytes read in buffer from _stream.
+        private int _writePos; // Write pointer within shared buffer.
+
+        private Task<int> _lastSyncCompletedReadTask; // The last successful Task returned from ReadAsync
+        // (perf optimization for successive reads of the same size)
+
+        private readonly SemaphoreSlim _sem = new SemaphoreSlim(1, 1);
+
+        public BufferedStream(Stream stream) : this(stream, DefaultBufferSize) {}
+
+
+        public BufferedStream(Stream stream, int bufferSize)
+        {
+            if (stream == null)
+                throw new ArgumentNullException("stream");
+
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException("bufferSize");
+
+            Contract.EndContractBlock();
+
+            Contract.Assert(!(stream is MemoryStream), "MemoryStream shouldn't be wrapped in a BufferedStream!");
+            Contract.Assert(!(stream is BufferedStream),
+                "BufferedStream shouldn't be wrapped in another BufferedStream!");
+
+            _stream = stream;
+            _bufferSize = bufferSize;
+
+            // Allocate _buffer on its first use - it will not be used if all reads
+            // & writes are greater than or equal to buffer size.
+
+            if (!_stream.CanRead && !_stream.CanWrite)
+                throw new ObjectDisposedException("BufferedStream");
+        }
+
+
+        private void EnsureNotClosed()
+        {
+            if (_stream == null)
+                throw new ObjectDisposedException("BufferedStream");
+        }
+
+
+        private void EnsureCanSeek()
+        {
+            Contract.Requires(_stream != null);
+
```

Comment:
I'll stop commenting on these, but think we should avoid Contract.Requires, etc.  Contract.Assert is probably our safest usage pattern for now.